llvm/lib/Target/Sparc/SparcInstrVIS.td

```diff
@@ -303,4 +303,10 @@ def : Pat<(i64 (mulhs i64:$lhs, i64:$rhs)),
       (SUBrr (UMULXHI $lhs, $rhs),
              (ADDrr (ANDrr (SRAXri $lhs, 63), $rhs),
                     (ANDrr (SRAXri $rhs, 63), $lhs)))>;
+
+def : Pat<(i64 (ctlz i64:$src)), (LZCNT $src)>;
+// 32-bit LZCNT.
+// The zero extension will leave us with 32 extra leading zeros,
+// so we need to compensate for it.
+def : Pat<(i32 (ctlz i32:$src)), (ADDri (LZCNT (SRLri $src, 0)), (i32 -32))>;
```

Comment:
It may make sense to `Promote` 32-bit ctlz. IIUC DAG type legalizer does the same expansion, but has a benefit of potentially optimizing `shr` and `add` with outer instructions during DAG combining phase.

Reply:
Probably because `CTLZ_ZERO_UNDEF` is expanded. It should be treated the same way as `CTLZ`.
(Note: Sparc's LCTNZ appears to have well-defined behavior for zero input, unlike equivalent instructions on most other targets. If the behavior was not defined, the instruction would naturally match to `CTLZ_ZERO_UNDEF`, and `CTLZ` would need to be expanded.)

Reply:
IIRC not all registers can be safely used on V8+ by full width. Can this be an issue?
Maybe just set it to libcall on V8+ to be on the safe side? This is what gcc appears to be doing.

Reply:
> This is what gcc appears to be doing.

I didn't pass `-mvis3` :man_facepalming: